runtime/runtime/src/lib.rs

```diff
@@ -1483,59 +1484,270 @@ impl Runtime {
-        /// compared to others (whatever reason may be for that.) Splitting into smaller chunks
-        /// allows the load to be distributed across threads more evenly and any tail latency
-        /// reduced due to the last chunk(s) being smaller.
-        const TARGET_CHUNKS_PER_THREAD: usize = 4;
-        let total = &mut processing_state.total;
         let apply_state = &mut processing_state.apply_state;
-        let state_update = &mut processing_state.state_update;
-        let tx_vec = signed_txs.into_nonexpired_transactions();
+        let state_update = &processing_state.state_update;
         let len = tx_vec.len();
-        let chunk_count_target = rayon::current_num_threads() * TARGET_CHUNKS_PER_THREAD;
-        let chunk_size =
-            (tx_vec.len() / chunk_count_target).clamp(MIN_CHUNK_SIZE, ValidBitmask::BITS as _);
+        const CHUNK_SIZE: usize = 32;
         let protocol_version = processing_state.protocol_version;
         let (valid_masks, (accounts, access_keys)) = rayon::join(
             || {
-                tx_vec
-                    .par_chunks(chunk_size)
-                    .map(|txs| {
-                        let mut valid_mask: ValidBitmask = 0;
-                        for (idx, tx) in txs.iter().enumerate() {
+                let tx_chunks = tx_vec.transactions.par_chunks(CHUNK_SIZE);
+                let validity_chunks =
+                    tx_vec.transaction_validity_check_passed.par_chunks(CHUNK_SIZE);
+                tx_chunks
+                    .zip(validity_chunks)
+                    .map(|(txs, validities)| {
+                        let mut valid_mask: [Option<InvalidTxError>; CHUNK_SIZE] =
+                            core::array::from_fn(|_| None);
+                        for (idx, (tx, is_valid)) in txs.iter().zip(validities.iter()).enumerate() {
                             let tx_hash = tx.hash();
-                            let v = validate_transaction(
-                                &apply_state.config,
-                                tx.clone(),
+                            if !is_valid {
+                                valid_mask[idx] = Some(InvalidTxError::Expired);
+                                continue;
+                            }
+                            valid_mask[idx] = if let Err(err) = validate_actions(
+                                &apply_state.config.wasm_config.limit_config,
+                                tx.transaction.actions(),
                                 protocol_version,
-                            );
-                            if let Err((err, _)) = v {
+                            ) {
                                 tracing::debug!(?tx_hash, ?err, "transaction invalid");
-                                continue;
+                                Some(InvalidTxError::ActionsValidation(err))
+                            } else {
+                                None
```

Comment:
In the process of refactoring/inlining this lost signature validation (originally hidden inside `ValidatedTransaction::new`.) This is the entire (and the only) reason why this parallelism logic exists at all.